torch/_overrides.py

```diff
@@ -0,0 +1,264 @@
+"""
+Python implementation of __torch_function__
+
+While most of the torch API and handling for __torch_function__ happens
+at the C++ level, some of the torch API is written in Python so we need
+python-level handling for __torch_function__ overrides as well. The main
+developer-facing functionality in this file is the
+torch_function_dispatch decorator. This function can be applied to
+python functions in the torch.functional module to enable
+__torch_function__ overrides for that function. See the examples in the
+docstrings for torch_function_dispatch for details.
+
+NOTE: heavily inspired by NumPy's ``__array_function__`` (see:
+https://github.com/pytorch/pytorch/issues/24015 and
+https://www.numpy.org/neps/nep-0018-array-function-protocol.html
+)
+
+"""
+
+import functools
+import textwrap
+from . import _six
+if _six.PY3:
+    from inspect import getfullargspec
+    import collections
+    ArgSpec = collections.namedtuple('ArgSpec', 'args varargs keywords defaults')
+
+    def getargspec(func):
+        spec = getfullargspec(func)
+        return ArgSpec(spec.args, spec.varargs, spec.varkw, spec.defaults)
+else:
+    from inspect import getargspec
+
+from .tensor import Tensor
+
+
+_TENSOR_ONLY = [Tensor]
+
+def _get_overloaded_types_and_args(relevant_args):
+    """Returns a list of arguments on which to call __torch_function__.
+
+    Parameters
+    ----------
+    relevant_args : iterable of array-like
+        Iterable of array-like arguments to check for __torch_function__
+        methods.
+
+    Returns
+    -------
+    overloaded_types : collection of types
+        Types of arguments from relevant_args with __torch_function__ methods.
+    overloaded_args : list
+        Arguments from relevant_args on which to call __torch_function__
+        methods, in the order in which they should be called.
```

Comment:
This is definitely the sort of thing that a few examples would make things a lot clearer.

Reply:
But in this case, just a link to wherever the algorithm is explained is probably good enough

Reply:
This also needs to be kept consistent with the C++ version so it's good to cross-reference between the two implementations here.

Reply:
Actually, this is related to the conversation about whether or not the decorator should be in C++. If the decorator is implemented in C++, you don't need to maintain two copies of the same thing, because you'd just reuse the C++ version in that case.

Reply:
That is true. 

To do that we'd need to figure out how to hook into the C++ dispatching machinery from Python. It's not clear to me how to do that right now.

Reply:
Yeah, you don't have to do it for this PR. But something to think about.